Randomizer - Shuffle Roll

TheLynk opened the pull request:
Add new setting randomizer "Shuffle Roll", shuffle ability do a roll for Link
Logic update for Tree and Break Crates

I am not against feedback or ideas for the custom model
And for All language Translation 
And for logic

Roll item Model is compatible texture mod

https://github.com/user-attachments/assets/afcfeb61-37ce-40c1-825f-6679ef88d4e4

<!--- section:artifacts:start -->
### Build Artifacts
<!--- section:artifacts:end -->

TheLynk committed:
Create AutoSyncFork.yml

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Off

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Delete .github/workflows/AutoSyncFork.yml

TheLynk committed:
Add Shuffle Roll (V1)

TODO New model for items

TheLynk committed:
Clang Format

TheLynk committed:
New Model For Roll

TheLynk committed:
Rename Custom Model in Code

TheLynk committed:
Add more translation

TheLynk committed:
Tweak Translation

serprex commented:
@Pepper0ni was thinking we could make this progressive (no roll, bonk KO, no bonk KO)

I've been hesitant here since figuring out what logic depends on rolling jumps will be tedious

For custom model, a dice should be straightforward, but maybe a bit too much lateral thinking. An arrow going in a circle should be enough

(edit: didn't see you already cobbled together a custom model, that works too)

serprex commented on soh/soh/Enhancements/game-interactor/vanilla-behavior/GIVanillaBehavior.h:
````diff
@@ -316,6 +316,15 @@ typedef enum {
+    // ```
+    // #### `args`
+    // - None
+
````
```suggestion
```

serprex commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -376,6 +378,8 @@ bool Logic::CanUse(RandomizerGet itemName) {
         case RG_EPONA:
             return IsAdult && CanUse(RG_EPONAS_SONG);
+        case RG_ROLL:
+            return IsChild && IsAdult;
```
```suggestion
```
just use HasItem

TheLynk replied:
return HasItem(RG_ROLL);

serprex commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -6301,11 +6301,17 @@ s32 func_8083BBA0(Player* this, PlayState* play) {
+                                       GET_PLAYER_ANIM(PLAYER_ANIMGROUP_landing_roll, this->modelAnimType),
+                                       1.25f * sWaterSpeedFactor);
+        gSaveContext.ship.stats.count[COUNT_ROLLS]++;
+    }
```
```suggestion
    if (GameInteractor_Should(VB_ROLL, true)) {
        Player_SetupAction(play, this, Player_Action_Roll, 0);
        LinkAnimation_PlayOnceSetSpeed(play, &this->skelAnime,
                                       GET_PLAYER_ANIM(PLAYER_ANIMGROUP_landing_roll, this->modelAnimType),
                                       1.25f * sWaterSpeedFactor);
        gSaveContext.ship.stats.count[COUNT_ROLLS]++;
    }
```
function is `void`, tho why here instead of `Player_TryRoll` which it seemed you were originally going for with this return?

TheLynk replied:
thanks

Pepper0ni commented:
This does sound like a good time to implement bonk damage logic, and can agree with wanting BONKO first then everything else to minimize navigation logic headaches. I think we will want to handle the various "basic movement" shuffles together as we can check for everything at the same time.

TheLynk committed:
Update soh/soh/Enhancements/game-interactor/vanilla-behavior/GIVanillaBehavior.h

Co-authored-by: Philip Dubé <serprex@users.noreply.github.com>

TheLynk committed:
Tweak code | Thx @serprex

TheLynk commented:
> This does sound like a good time to implement bonk damage logic, and can agree with wanting BONKO first then everything else to minimize navigation logic headaches. I think we will want to handle the various "basic movement" shuffles together as we can check for everything at the same time.

so this would have to be done on another branch and by who?

serprex commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -376,6 +378,8 @@ bool Logic::CanUse(RandomizerGet itemName) {
             return IsAdult && CanUse(RG_EPONAS_SONG);
+        case RG_ROLL:
+            return HasItem(RG_ROLL);
 
```
```suggestion
```

No need to handle this, CanUse returns false when `HasItem` is false, & we have an assertion to avoid using `CanUse` with items that should only use `HasItem` (I think it's a bit overly complicated, that it'd be nice for logic to consolidate on always using CanUse, but that'd be outside this PR)

TheLynk committed:
Update soh/soh/Enhancements/randomizer/logic.cpp

Co-authored-by: Philip Dubé <serprex@users.noreply.github.com>

TheLynk committed:
Fix translation

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Re Add All the branch (but no Shuffle Silver Rupee) | V1.0

TheLynk committed:
Add Branch "Shuffle Roll" | Version : 2.0

TheLynk committed:
Fix Text in Item tracker and Logic for ZR | V:2.1

TheLynk committed:
Fix Logic in KF

TheLynk committed:
Fix Logic in KF | 2.2

TheLynk committed:
Add Branch add the ability to Randomize Music (Seeded) | V:3.0

TheLynk committed:
Add branch "Invisible Navi option" | V:4.0

TheLynk committed:
Add branch "Rando: Add Triforce Hunt GBK setting" | V:5.0

TheLynk committed:
Add branch "Add Edit buttons for custom presets" | V6.0

TheLynk committed:
Add branch "Check tracker improvements" | V:7.0

TheLynk committed:
Add branch "Improvements to Custom Kaleido Menu" | V:8.0

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Fix Logic

TheLynk committed:
Fix Logic | V:8.1

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Fix Logic Shuffle Bean Souls | V:8.2

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'develop' into Shuffle-Roll

TheLynk committed:
Update model for Shuffle Roll | V:8.3

TheLynk committed:
Fix Merge And New Model for Roll

TheLynk commented:
from memory I think I have a line that is no longer useful in draw.cpp that I can remove but I will do that later

TheLynk committed:
Merge branch 'develop' into V3-Anchor-Fusion-All-New-Option

TheLynk committed:
Fix Update Auto Merge | V:9.0

TheLynk committed:
Fix crash Logic | V:9.1

TheLynk committed:
Add branch "rocksanity" | V:10.0

TheLynk committed:
clang-format

TheLynk committed:
Fix Logic | V:10.1

serprex reviewed (approved):
#5987 is an example of trick using rolling jump, but I've been paying attention lately & I don't see rolling jumps being base logic. I think we can merge this as is. There is the question of tricks involving setups that involve rolls/siderolls. But okay for that to be out of scope

serprex commented on soh/soh/Enhancements/randomizer/draw.cpp:
```diff
@@ -1109,6 +1113,48 @@ extern "C" void Randomizer_DrawBronzeScale(PlayState* play, GetItemEntry* getIte
+                           ROLL_ANIMATION_LIMBS);
+    }
+
+    // Animation manuelle avec les frames capturées
```
comments should be English

TheLynk replied:
Oups sorry

TheLynk replied:
Fix

serprex commented on soh/soh/Enhancements/randomizer/roll_animation_data.h:
```diff
@@ -0,0 +1,348 @@
+#define ROLL_ANIMATION_FRAMES 20
+#define ROLL_ANIMATION_LIMBS 22
+
+static const Vec3s rollAnimationData[ROLL_ANIMATION_FRAMES][ROLL_ANIMATION_LIMBS] = {
```
where did this data come from?

TheLynk replied:
I had done a piece of work to capture the animation of link when he rolls and then I retrieved the results of capturing the data to use them for the model

serprex replied:
It'd be useful to have that made available somewhere so we can apply it for other shuffle items

TheLynk replied:
Okay, well I'll put it in a separate file so the other developers can use it.

TheLynk replied:
@serprex it's good I added animation_capture_tool.cpp this is the code that I used to make my roll model and I even translated the comment into English and I also left it in French

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

PABessero committed:
Merge branch 'develop' into V3-Anchor-Fusion-All-New-Option

good luck

TheLynk committed:
Major Update | V:11

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'develop' into V3-Anchor-Fusion-All-New-Option

TheLynk committed:
Good Luck Update | V:12

small issue is found is if open chest is not shuffle the randomizer run is imbattable

TheLynk committed:
Merge remote-tracking branch 'upstream/develop' into Shuffle-Roll

Part 1

TheLynk committed:
Merge remote-tracking branch 'upstream/develop' into Shuffle-Roll

Part 2

TheLynk committed:
Add Animation Capture Tool.cpp & remove useless code in draw.cpp

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Update animation_capture_tool.cpp

serprex commented on soh/soh/Enhancements/randomizer/animation_capture_tool.cpp:
```diff
@@ -0,0 +1,57 @@
+
+// and in this example I had temporarily put it in the Draw.cpp file and I called the function when Link performed a
+// roll et dans cet exemple je l'avais mis temporairement dans le fichier Draw.cpp et j'appelais la fonction au moment
+// que Link fessait une roulade
```
```suggestion
// I provide here the code which allowed me to have all the data for the roll animation for roll_animation_data.h
// I had temporarily put it in the Draw.cpp file and I called the function when Link performed a roll
```

serprex commented:
> > This does sound like a good time to implement bonk damage logic, and can agree with wanting BONKO first then everything else to minimize navigation logic headaches. I think we will want to handle the various "basic movement" shuffles together as we can check for everything at the same time.
> 
> so this would have to be done on another branch and by who?

@Pepper0ni has said he'd going over logic, will also be noting other basic movement when he does sweep. But is somewhat held up atm

Pepper0ni commented:
Just to talk about my worries with logic, on top of the obvious stuff with rolljumps (which are usually tricks, but there could be oversights) one thing I would double check is hover boots movement, as rolling in hovers is often logical and if there's a ledge grab at the end backwalking may not replace it without a backflip (which would be a trick) 

I am indeed fighting through my todo list ATM and now serp is merging things I have plenty to review, so I can't guarantee a swift response. If we want to merge before my sweep a beta period to find those oversights before the next release is advised

TheLynk commented:
> Just to talk about my worries with logic, on top of the obvious stuff with rolljumps (which are usually tricks, but there could be oversights) one thing I would double check is hover boots movement, as rolling in hovers is often logical and if there's a ledge grab at the end backwalking may not replace it without a backflip (which would be a trick) 
> 
> I am indeed fighting through my todo list ATM and now serp is merging things I have plenty to review, so I can't guarantee a swift response. If we want to merge before my sweep a beta period to find those oversights before the next release is advised

I'm already running a beta test on another branch, which itself has several branches that are also in testing.

TheLynk committed:
Update soh/soh/Enhancements/randomizer/animation_capture_tool.cpp

Co-authored-by: Philip Dubé <serprex@users.noreply.github.com>

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'develop' into Shuffle-Roll

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Clang Format

TheLynk committed:
Fix crash and Add Ice traps Text for Roll

serprex commented:
see #6184 for integrating into kaleido

TheLynk commented:
> see #6184 for integrating into kaleido

ok I will make the modification but I don't know how to get it Kaleido Tracker in game

serprex commented:
on quest screen in pause menu hit C-up

TheLynk committed:
Add Roll in Kaleido Tracker

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop

TheLynk committed:
Merge branch 'HarbourMasters:develop' into develop (Manuel)

TheLynk committed:
Add Roll in Item Tracker

TheLynk committed:
To many addition | V:13

TheLynk committed:
To many addition | V:13

TheLynk committed:
Update Logic | V:13.1

TheLynk committed:
Merge branch 'V3-Anchor-Fusion-All-New-Option' into Shuffle-Roll

TheLynk commented:
Fuck bad handling

TheLynk committed:
Revert "Merge branch 'V3-Anchor-Fusion-All-New-Option' into Shuffle-Roll"

This reverts commit a8fee8ed4acc89444c6a102908c0dd06b198e56a, reversing
changes made to 1f2c3832762ae2a50683c781e1a03958ac648dc7.

TheLynk commented:
It's okay, I repaired and restored everything (I'm really a drag)

TheLynk commented:
~~bug only on linux build : https://discord.com/channels/808039310850130000/1389100381161783516/1465124219221442716~~
Fix

TheLynk committed:
Add Missing Line

Pepper0ni commented:
I would like to get around to pushing this more, but this needs the PR i sent merging (unless you want me to repeat myself in the PR review) and the branch updating

Malkierian commented:
Which other PR? Where was this mentioned?

Pepper0ni commented:
I sent it to their fork https://github.com/TheLynk/Shipwright/pull/16 basically me doing a review directly with changes.

TheLynk commented:
I will soon be able to resume work on this project because I have almost finished what I wanted to do on the pikmin 1 apworld where I am the main dev and if all the tests go well in max 2 weeks I could work on the Shuffle Roll update

Pepper0ni commented:
Given how long this has sat with no activity, if there's no obvious activity before shuffle silvers merges, I'm going to take responsibility for getting this up to date and merge ready once that happens.

Having such a major shuffle stuck in the pipe is only going to cause headaches for future planning with tricks so I want this sorted.

TheLynk commented:
I had completely forgotten that I had to completely update this branch, I'll get started on it this week

Pepper0ni commented:
Remember the PR i sent with those logic fixes too

TheLynk committed:
Merge remote-tracking branch 'upstream/develop' into Shuffle-Roll

TheLynk committed:
Revert "Merge remote-tracking branch 'upstream/develop' into Shuffle-Roll"

This reverts commit f56452d10d35fed3a5968fc8b558a1955aeaadad, reversing
changes made to 447272879a099437a18424e359f0c673cb88d15d.
